src/box/applier.cc

```diff
@@ -1252,6 +1341,551 @@ applier_on_rollback(struct trigger *trigger, void *event)
+		 *
+		 * TODO: let's find a way to pin data on the ibuf so that it is
+		 * never relocated while it is needed. We'll then get rid of
+		 * copying row bodies altogether.
+		 * We'll probably have to invent some kind of a new net buffer
+		 * to perform this task.
+		 * Looks like lsregion does almost exactly what we need, maybe
+		 * there's a way to use it instead of ibuf?
+		 */
+		size_t len = row->body->iov_len;
+		void *new_base = lsregion_alloc(lsr, len,
+						++applier->thread.lsr_id);
+		if (new_base == NULL)
+			tnt_raise(OutOfMemory, len, "lsregion_alloc",
+				  "xrow body");
+		memcpy(new_base, row->body->iov_base, row->body->iov_len);
+		/* Adjust row body pointers. */
+		row->body->iov_base = new_base;
+	}
+}
+
+/** Applier thread reader fiber function. */
+static int
+applier_thread_reader_f(va_list ap)
+{
+	struct applier *applier = va_arg(ap, struct applier *);
+	struct applier_thread *thread = container_of(cord(), typeof(*thread),
+						     cord);
+	struct lsregion *lsr = &applier->thread.lsr;
+	const struct applier_read_ctx ctx = {
+		.ibuf = &applier->thread.ibuf,
+		.alloc_row = thread_alloc_row,
+		.save_body = thread_save_body,
+	};
+
+	while (!fiber_is_cancelled()) {
+		double timeout = applier->version_id < version_id(1, 7, 7) ?
+				 TIMEOUT_INFINITY :
+				 replication_disconnect_timeout();
+		struct applier_tx *tx;
+		tx = lsregion_alloc_object(lsr, applier->thread.lsr_id++,
+					   struct applier_tx);
+		if (tx == NULL) {
+			diag_set(OutOfMemory, sizeof(*tx),
+				 "lsregion_alloc_object", "tx");
+			goto exit_notify;
+		}
+		try {
+			applier_read_tx(applier, &tx->rows, &ctx, timeout);
+		} catch (FiberIsCancelled *) {
+			return 0;
+		} catch (Exception *e) {
+			goto exit_notify;
+		}
+		struct applier_data_msg *msg;
+		do {
+			msg = applier_thread_next_msg(applier);
+			if (msg != NULL)
+				break;
+			fiber_yield();
+			if (fiber_is_cancelled())
+				return 0;
+		} while (true);
+		applier_thread_push_tx(thread, msg, tx);
+	}
+	return 0;
+exit_notify:
+	/* Notify the tx thread that its applier exited with an error. */
+	assert(!diag_is_empty(diag_get()));
+	diag_move(diag_get(), &applier->thread.exit_msg.diag);
+	cpipe_push(&thread->tx_pipe, &applier->thread.exit_msg.base.base);
+	return 0;
+}
+
+/** The main applier thread fiber function. */
+static int
+applier_thread_f(va_list ap)
+{
+	struct applier_thread *thread = va_arg(ap, typeof(thread));
+	int rc = cbus_endpoint_create(&thread->endpoint, cord()->name,
+				      fiber_schedule_cb, fiber());
+	assert(rc == 0);
+	(void)rc;
+
+	cpipe_create(&thread->tx_pipe, "tx_prio");
+
+	cbus_loop(&thread->endpoint);
+
+	unreachable();
+}
+
+/** Initialize and start the applier thread. */
+static void
+applier_thread_create(struct applier_thread *thread)
+{
+	static int thread_id = 0;
+	const char *name = tt_sprintf("applier_%d", ++thread_id);
+
+	memset(thread, 0, sizeof(*thread));
+
+	if (cord_costart(&thread->cord, name, applier_thread_f, thread) != 0)
```

Comment:
I see costart and see that the threads can gracefully end. But where is `cord_cojoin/join` then?

Reply:
I must've misunderstood @locker, when he said that we may create the threads at startup and never start/stop them dynamically.

I sure can add `applier_free()` which will stop and join all the applier threads.

Reply:
I think it'd be fine to cancel and join them with `pthread_cancel`, like we do with iproto threads.

Reply:
Done:

<details><summary>diff</summary>

```diff
diff --git a/src/box/applier.cc b/src/box/applier.cc
index 7d4a5518e..5421fb686 100644
--- a/src/box/applier.cc
+++ b/src/box/applier.cc
@@ -1689,6 +1689,18 @@ applier_init(void)
        }
 }
 
+void
+applier_free(void)
+{
+       for (int i = 0; i < replication_threads; i++) {
+               struct applier_thread *thread = applier_threads[i];
+               tt_pthread_cancel(thread->cord.id);
+               tt_pthread_join(thread->cord.id, NULL);
+               free(thread);
+       }
+       free(applier_threads);
+}
+
 /** Get a working applier thread. */
 static struct applier_thread *
 applier_thread_next(void)
diff --git a/src/box/applier.h b/src/box/applier.h
index 3e3c1308d..dd311a873 100644
--- a/src/box/applier.h
+++ b/src/box/applier.h
@@ -216,6 +216,10 @@ struct applier {
 void
 applier_init(void);
 
+/** Free the applier subsystem. */
+void
+applier_free(void);
+
 /**
  * Start a client to a remote master using a background fiber.
  *
diff --git a/src/box/replication.cc b/src/box/replication.cc
index 8bd45e41e..bbf4b7417 100644
--- a/src/box/replication.cc
+++ b/src/box/replication.cc
@@ -121,6 +121,8 @@ replication_free(void)
 
        diag_destroy(&replicaset.applier.diag);
        trigger_destroy(&replicaset.on_ack);
+
+       applier_free();
 }
 
 int
```
</details>